src/server/scripts/Northrend/Ulduar/Ulduar/boss_xt002.cpp

```diff
@@ -137,960 +112,865 @@ enum XT002Data
-                    {
-                        DoCastVictim(SPELL_TRAMPLE);
-                        _trampleTimer = TIMER_TRAMPLE;
-                    }
-                    else
-                        _trampleTimer -= diff;
-
-                    if (_uppercutTimer <= diff)
-                    {
-                        DoCastVictim(SPELL_UPPERCUT);
-                        _uppercutTimer = TIMER_UPPERCUT;
-                    }
-                    else
-                        _uppercutTimer -= diff;
-                }
+                me->SetReactState(REACT_AGGRESSIVE);
+                DoZoneInCombat();
+            }).Schedule(Seconds(17), [this](TaskContext trample)
+            {
+                DoCastSelf(SPELL_TRAMPLE);
+                trample.Repeat(11s);
+            }).Schedule(Seconds(19), [this](TaskContext arcingSmash)
+            {
+                DoCastSelf(SPELL_ARCING_SMASH);
+                arcingSmash.Repeat(8s);
+            }).Schedule(Seconds(19), [this](TaskContext upperCut)
+            {
+                DoCastVictim(SPELL_UPPERCUT);
+                upperCut.Repeat(14s);
+            });
 
-                DoMeleeAttackIfReady();
-            }
+    }
 
-            private:
-                InstanceScript* _instance;
-                uint32 _arcingSmashTimer;
-                uint32 _trampleTimer;
-                uint32 _uppercutTimer;
-        };
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
 
-        CreatureAI* GetAI(Creature* creature) const override
+        _scheduler.Update(diff, [this]
         {
-            return GetUlduarAI<npc_pummellerAI>(creature);
-        }
+            DoMeleeAttackIfReady();
+        });
+    }
+
+private:
+    InstanceScript* _instance;
+    TaskScheduler _scheduler;
 };
 
-/*-------------------------------------------------------
- *
- *        XE-321 BOOMBOT
- *
- *///----------------------------------------------------
 class BoomEvent : public BasicEvent
 {
-    public:
-        BoomEvent(Creature* me) : _me(me)
-        {
-        }
+public:
+    BoomEvent(Creature* me) : _me(me) { }
 
-        bool Execute(uint64 /*time*/, uint32 /*diff*/) override
-        {
-            // This hack is here because we suspect our implementation of spell effect execution on targets
-            // is done in the wrong order. We suspect that EFFECT_0 needs to be applied on all targets,
-            // then EFFECT_1, etc - instead of applying each effect on target1, then target2, etc.
-            // The above situation causes the visual for this spell to be bugged, so we remove the instakill
-            // effect and implement a script hack for that.
-
-            _me->CastSpell(_me, SPELL_BOOM, false);
-            return true;
-        }
+    bool Execute(uint64 /*time*/, uint32 /*diff*/) override
+    {
+        // This hack is here because we suspect our implementation of spell effect execution on targets
```

Comment:
This part will be fixed in https://github.com/TrinityCore/TrinityCore/pull/21395